protocol/rpcconsumer/consumer_relay_state_machine.go

```diff
@@ -83,24 +111,72 @@ func (crsm *ConsumerRelayStateMachine) GetSelection() Selection {
 	return crsm.selection
 }
 
-func (crsm *ConsumerRelayStateMachine) shouldRetryOnResult(numberOfRetriesLaunched int, numberOfNodeErrors uint64) bool {
-	shouldRetry := crsm.shouldRetryInner(numberOfRetriesLaunched)
-	// archive functionality will be added here.
+// Should retry implements the logic for when to send another relay.
+// As well as the decision of changing the protocol message,
+// into different extensions or addons based on certain conditions
+func (crsm *ConsumerRelayStateMachine) shouldRetry(numberOfRetriesLaunched int, numberOfNodeErrors uint64) bool {
+	shouldRetry := crsm.retryCondition(numberOfRetriesLaunched)
+	if shouldRetry {
+		// Retry archive logic
+		hashes := crsm.GetProtocolMessage().GetRequestedBlocksHashes()
+		if len(hashes) > 0 && numberOfNodeErrors > 0 {
+			// Launch archive only on the second retry attempt.
+			if numberOfRetriesLaunched == 1 {
+				// Iterate over all hashes found in relay, if we don't have them in the cache we can try retry on archive.
+				// If we are familiar with all, we don't want to allow archive.
+				for _, hash := range hashes {
+					if !crsm.relayRetriesManager.CheckHashInCache(hash) {
+						// If we didn't find the hash in the cache we can try archive relay.
+						relayRequestData := crsm.protocolMessage.RelayPrivateData()
+						// Validate we're not already archive
+						if slices.Contains(relayRequestData.Extensions, extensionslib.ArchiveExtension) {
+							break // Do nothing its already archive.
+						}
+						// We need to set archive.
+						// Create a new relay private data containing the extension.
+						userData := crsm.protocolMessage.GetUserData()
+						metaDataForArchive := []pairingtypes.Metadata{{Name: common.EXTENSION_OVERRIDE_HEADER_NAME, Value: extensionslib.ArchiveExtension}}
+						newProtocolMessage, err := crsm.relaySender.ParseRelay(crsm.ctx, relayRequestData.ApiUrl, string(relayRequestData.Data), relayRequestData.ConnectionType, userData.DappId, userData.ConsumerIp, metaDataForArchive)
+						if err != nil {
+							utils.LavaFormatError("Failed converting to archive message in shouldRetry", err, utils.LogAttr("relayRequestData", relayRequestData), utils.LogAttr("metadata", metaDataForArchive))
+						}
+						// Creating an archive protocol message, and set it to current protocol message
+						crsm.protocolMessage = newProtocolMessage
```

Comment:
you shouldnt overwrite the protocol message, the class shouldnt even hold it anymore, it should be within the states objects